pkg/driver/mount.go

```diff
@@ -17,24 +17,317 @@ limitations under the License.
+			return instance, nil
+		} else if retry <= MaxRetryCount {
+			sleepInterval = 5
+			if instance.Status.Status == "RO" {
+				replicaStatus := instance.Status.ReplicaStatuses
+				if len(replicaStatus) != 0 {
+					logrus.Warningf("Volume: {%v} is in RO mode: replica status: {%+v}", volID, replicaStatus)
+					continue
+				}
+				logrus.Warningf("Volume: {%v} is not ready: replicas may not be connected", volID)
+				continue
+			}
+			logrus.Warningf("Volume: {%v} is not ready: volume status is {%s}", volID, instance.Status.Status)
+			continue
+		} else {
+			break
+		}
+	}
+	return nil, fmt.Errorf("Max retry count exceeded, volume: {%v} is not ready", volID)
+}
+
+func waitForVolumeToBeReachable(targetPortal string) error {
+	var (
+		retries int
+		err     error
+		conn    net.Conn
+	)
+
+	for {
+		// Create a connection to test if the iSCSI Portal is reachable,
+		if conn, err = net.Dial("tcp", targetPortal); err == nil {
+			conn.Close()
+			logrus.Debugf("Target: {%v} is reachable to create connections", targetPortal)
+			return nil
+		}
+		// wait until the iSCSI targetPortal is reachable
+		// There is no pointn of triggering iSCSIadm login commands
+		// until the portal is reachable
+		time.Sleep(2 * time.Second)
+		retries++
+		if retries >= MaxRetryCount {
+			// Let the caller function decide further if the volume is
+			// not reachable even after 12 seconds ( This number was arrived at
+			// based on the kubelets retrying logic. Kubelet retries to publish
+			// volume after every 14s )
+			return fmt.Errorf(
+				"iSCSI Target not reachable, TargetPortal: {%v}, err: {%v}",
+				targetPortal, err)
+		}
+	}
+}
+
+func listContains(
+	mountPath string, list []mount.MountPoint,
+) (*mount.MountPoint, bool) {
+	for _, info := range list {
+		if info.Path == mountPath {
+			mntInfo := info
+			return &mntInfo, true
+		}
+	}
+	return nil, false
+}
+
+// MonitorMounts makes sure that all the volumes present in the inmemory list
+// with the driver are mounted with the original mount options
+// This function runs a never ending loop therefore should be run as a goroutine
+// Mounted list is fetched from the OS and the state of all the volumes is
+// reverified after every 5 seconds. If the mountpoint is not present in the
+// list or if it has been remounted with a different mount option by the OS, the
+// volume is added to the ReqMountList which is removed as soon as the remount
+// operation on the volume is complete
+// For each remount operation a new goroutine is created, so that if multiple
+// volumes have lost their original state they can all be remounted in parallel
+func (n *NodeMounter) MonitorMounts() {
+	logrus.Infof("Starting MonitorMounts goroutine")
+	var (
+		err        error
+		csivolList *jv.JivaVolumeList
+		mountList  []mount.MountPoint
+	)
+	ticker := time.NewTicker(MonitorMountRetryTimeout * time.Second)
+	for {
+		select {
+		case <-ticker.C:
```

Comment:
Please take a lock over here for the whole jivaVolumeList and release it only when all the RO volumes have been added to transition list. Please go through the same function in cstor-csi https://github.com/openebs/cstor-csi/blob/master/pkg/utils/v1alpha1/utils.go#L213